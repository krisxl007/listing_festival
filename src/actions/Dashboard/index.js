import { FETCH_DASHBOARD } from '../../constants/ActionType';
import { GET } from '../../utils/Api';
import { ERROR, WARNING } from '../../constants/Toast';

export const fetchDashboardData = (token, addToast) => dispatch => {

  const url = 'http://eacodingtest.digital.energyaustralia.com.au/api/v1/festivals';
  const header = {
    headers: {
      Authorization: "Bearer " + token,
      Accept: "application/json, text/plain, */*",
      "Content-Type": "application/json"
    }
  };

  GET(url, header).then(response => {
    if(response.data === "") {
      addToast('No festival data found, please fetch it again.', WARNING)
    }
    
    dispatch({
      type: FETCH_DASHBOARD,
      payload: {
        data: response.data
      }
    });
  }).catch(error => {
    if(error.response) {
      addToast(error.response.data, ERROR)
    }else {
      addToast(error.message, ERROR)
    }
    
    dispatch({
      type: FETCH_DASHBOARD,
      payload: {
        data: ''
      }
    });
  })

};
