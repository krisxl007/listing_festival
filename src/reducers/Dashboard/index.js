import { FETCH_DASHBOARD } from '../../constants/ActionType';

const initialState = {
  dashboardData: []
};

export default function(state = initialState, action) {
  switch (action.type) {
    case FETCH_DASHBOARD: {
      let data = [];
      if(action.payload && action.payload.data) {
        data = action.payload.data;
      }

      return Object.assign({}, state, { dashboardData: data });
    }
    default: {
      return state;
    }
  }
}
