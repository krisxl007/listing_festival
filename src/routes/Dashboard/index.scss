.dashboard__welcome-div {
  float: left;
  padding: 20px;
  font-weight: bold;
}

.dashboard__logout-div {
  float: right;
  padding: 20px;
}

.dashboard__content-div {
  padding-top: 100px;
  padding-left: 20px;
}

.dashboard__fetch-festival-list-div {
  float: right;
  padding-right: 20px;

}

.dashboard__festival-list-div {
  font-weight: bold;
}

.dashboard__festival-list-label-div {
  background-color: yellow;
}

.dashboard__festival-list-band-div {
  padding-left: 20px;
}

.dashboard__festival-list-festival-div {
  padding-left: 40px;
}
