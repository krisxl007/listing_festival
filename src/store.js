import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import dashboard from './reducers/Dashboard'

// Application reducers

// All application reducer combine definitions
const reducers = combineReducers({
  dashboard
});

// Default redux middleware definitions
const middlewareList = [thunk];

// We only want redux state logging in development mode
if (process.env.NODE_ENV === 'development') {
  middlewareList.push(logger);
}

const middleware = applyMiddleware(...middlewareList);

export default createStore(reducers, middleware);
