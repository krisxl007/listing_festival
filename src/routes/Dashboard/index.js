import React, { useEffect } from 'react';
import { withRouter } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { fetchDashboardData } from '../../actions/Dashboard';
import tokenValidator from '../../utils/Validator';
import Logout from '../../components/Logout';
import { useToasts } from 'react-toast-notifications';

import './index.scss';

export function Dashboard({ history }) {
  const { addToast } = useToasts();
  const dispatch = useDispatch();
  const data = useSelector(state => state.dashboard);

  useEffect(() => {
    const token = tokenValidator('token', history, '/');
    if (token != null) {
      dispatch(fetchDashboardData(token, addToast));
    }
  }, [history, dispatch]);

  const constructData = () => {
    const dashBoardData = data.dashboardData;
    console.log('dashBoardData', dashBoardData);

    let labelList = [];
    for (let item of dashBoardData) {
      if (item.bands && item.bands.length > 0) {
        for (let band of item.bands) {
          if (band.recordLabel) {
            addLabel(labelList, band, band.recordLabel, item);
          } else {
            // if no record label, put to unknown
            addLabel(labelList, band, 'Unknown', item);
          }
        }
      }
    }

    console.log('labelList', labelList);

    return (
      <div>
        {labelList.map(each => {
          return (
            <div>
              <br />
              <label className="dashboard__festival-list-label-div">
                {each.labelName}
              </label>
              {each.bands &&
                each.bands.map(band => {
                  return (
                    <div className="dashboard__festival-list-band-div">
                      <div>- {band.bandName}:</div>
                      {band.festivalList &&
                        band.festivalList.map(festival => {
                          return (
                            <div className="dashboard__festival-list-festival-div">
                              - {festival}
                            </div>
                          );
                        })}
                    </div>
                  );
                })}
            </div>
          );
        })}
      </div>
    );
  };

  const addLabel = (labelList, band, labelName, festivalItem) => {
    const label = labelList.find(label => label.labelName === labelName);
    if (label && label.bands) {
      const labelBand = label.bands.find(each => each.bandName === band.name);
      if (labelBand) {
        labelBand.festivalList.push(
          festivalItem.name ? festivalItem.name : 'Unknown'
        );
        labelBand.festivalList.sort(); // sort festival order
      } else {
        // if band not found, add a new record to it
        label.bands.push({
          bandName: band.name,
          festivalList: [festivalItem.name ? festivalItem.name : 'Unknown']
        });
      }
    } else {
      // if label not found, add a new record to it
      labelList.push({
        labelName: labelName,
        bands: [
          {
            bandName: band.name,
            festivalList: [festivalItem.name ? festivalItem.name : 'Unknown']
          }
        ]
      });
    }
  };

  const fetchFestival = () => {
    const token = sessionStorage.getItem('token');
    dispatch(fetchDashboardData(token, addToast));
  };

  const displayData = () => {
    return (
      <div>
        <div className="dashboard__fetch-festival-list-div">
          <button onClick={fetchFestival}>Fetch Festival List</button>
        </div>
        <div>
          <div className="dashboard__festival-list-div">Festival List:</div>
          <div>{constructData()}</div>
        </div>
      </div>
    );
  };

  return (
    <div>
      <div>
        <div className="dashboard__welcome-div">Welcome to dashboard!</div>
        <div className="dashboard__logout-div">
          <Logout />
        </div>
      </div>
      <div className="dashboard__content-div">{displayData()}</div>
    </div>
  );
}

export default withRouter(Dashboard);
