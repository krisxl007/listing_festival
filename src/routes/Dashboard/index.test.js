import React from 'react';
import { mount, shallow } from 'enzyme';
import Dashboard from './index';
import { config } from '../../utils/Enzyme/enzymeConfig';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import store from '../../store';
import thunk from 'redux-thunk';

describe('Test Dashboard component', () => {
  beforeAll(() => {
    sessionStorage.setItem('authenticated', 'xxxxx');
  });

  afterAll(() => {
    sessionStorage.removeItem('authenticated');
  });

  it('snapshot test', () => {
    const initialState = {
      dashboardData: {
        data: 'aaa'
      }
    };
    const mockStore = configureStore();

    let wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>
    );

    expect(wrapper.debug()).toContain('dashboard data');
    expect(wrapper).toMatchSnapshot();
  });

});
