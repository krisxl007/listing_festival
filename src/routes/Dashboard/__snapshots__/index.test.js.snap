// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dashboard component on mount snapshot test 1`] = `ShallowWrapper {}`;

exports[`Test Dashboard component snapshot test 1`] = `ReactWrapper {}`;
